File: math_algrithm_book/009.py
N, S = map(int, input().split())
list_a = list(map(int, input().split()))

# ビット全探索
# 0-2^Nまでループ
# 1<<n は，2^nの意味
for bit in range(1 << N):
    sum = 0
    for i in range(N):
        # bit=5, i<<1なら，5(101)の2^1bit目とAND比較する．
        if bit & (1 << i):
            # i番目のカードが選ばれている場合、その値を合計に加える
            sum += list_a[i]

    # 合計がSと一致するか確認
    if sum == S:
        print("Yes")
        break
else:
    print("No")

File: math_algrithm_book/chapter2/009.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
    int N, S;
    cin >> N >> S;
    vector<int> A(N);
    for (int i = 0; i < N; i++)
    {
        cin >> A[i];
    }

    bool found = false;

    // 全ての部分集合を試す
    for (int bit = 0; bit < (1 << N); bit++) // bit は 0 から 7 までの値を取る
    {
        int sum = 0;
        for (int i = 0; i < N; i++) // N=3 なので、i は 0 から 2 までの値を取る
        {
            if (bit & (1 << i)) // bit の i 番目のビットが 1 ならば
            {
                sum += A[i]; // sum に A[i] を加える
            }
        }
        if (sum == S) // 合計が S に一致するかをチェック
        {
            found = true;
            break;
        }
    }

    if (found)
    {
        cout << "Yes" << endl;
    }
    else
    {
        cout << "No" << endl;
    }

    return 0;
}
